akka-remote/src/main/scala/akka/remote/Remoting.scala

```diff
@@ -548,18 +555,28 @@ private[remote] class EndpointManager(conf: Config, log: LoggingAdapter) extends
             "address cannot be quarantined without knowing the UID, gating instead for {} ms.",
             address, settings.RetryGateClosedFor.toMillis)
           endpoints.markAsFailed(endpoint, Deadline.now + settings.RetryGateClosedFor)
-        case (Some(Pass(endpoint, Some(currentUid), _)), Some(quarantineUid)) if currentUid == quarantineUid ⇒
-          context.stop(endpoint)
-        case _ ⇒
-        // Do nothing, because either:
-        // A: we don't know yet the UID of the writer, it will be checked against current quarantine state later
-        // B: we know the UID, but it does not match with the UID to be quarantined
+        case (Some(Pass(endpoint, uidOption, refuseUidOption)), Some(quarantineUid)) ⇒
+          uidOption match {
+            case Some(uid) if uid == quarantineUid ⇒
+              endpoints.markAsQuarantined(address, quarantineUid, Deadline.now + settings.QuarantineDuration)
+              eventPublisher.notifyListeners(QuarantinedEvent(address, quarantineUid))
+              context.stop(endpoint)
+            // or it does not match with the UID to be quarantined
+            case None if !refuseUidOption.contains(quarantineUid) ⇒
+              // the quarantine uid may be got fresh by cluster gossip, so update refuseUid for late handle when the writer got uid
+              endpoints.registerWritableEndpointRefuseUid(address, quarantineUid)
```

Comment:
the refuseUid checked by writer comes from the constructor parameter. so the above code has no affect on it. actually, we don't distinguish the writer(bottom) and cluster(top) 's quarantineUid which one is the newest. so the double-check needed in some time.
If the race happened, the manager will receive the ```GotUid``` msg, then it will recheck it against what's the ```endpoints``` remember(which may be from the cluster).